health/health_log.c

```diff
@@ -358,54 +361,100 @@ inline void health_alarm_log(
-    ae->old_status = old_status;
-    ae->new_status = new_status;
-    ae->duration = duration;
-    ae->delay = delay;
-    ae->delay_up_to_timestamp = when + delay;
-    ae->flags |= flags;
+        if(chart) {
+            ae->chart = strdupz(chart);
+            ae->hash_chart = simple_hash(ae->chart);
+        }
 
-    if(ae->old_status == RRDCALC_STATUS_WARNING || ae->old_status == RRDCALC_STATUS_CRITICAL)
-        ae->non_clear_duration += ae->duration;
+        if(family)
+            ae->family = strdupz(family);
+
+        if(exec) ae->exec = strdupz(exec);
+        if(recipient) ae->recipient = strdupz(recipient);
+        if(source) ae->source = strdupz(source);
+        if(units) ae->units = strdupz(units);
+        if(info) ae->info = strdupz(info);
+
+        ae->unique_id = host->health_log.next_log_id++;
+        ae->alarm_id = alarm_id;
+        ae->alarm_event_id = alarm_event_id;
+        ae->when = when;
+        ae->old_value = old_value;
+        ae->new_value = new_value;
+
+        char value_string[100 + 1];
+        ae->old_value_string = strdupz(format_value_and_unit(value_string, 100, ae->old_value, ae->units, -1));
+        ae->new_value_string = strdupz(format_value_and_unit(value_string, 100, ae->new_value, ae->units, -1));
+
+        ae->old_status = old_status;
+        ae->new_status = new_status;
+        ae->duration = duration;
+        ae->delay = delay;
+        ae->delay_up_to_timestamp = when + delay;
+        ae->flags |= flags;
+
+        ae->warn_repeat_every = warn_repeat_every;
+        ae->crit_repeat_every = crit_repeat_every;
+        ae->last_repeat = 0;
+
+        if(ae->old_status == RRDCALC_STATUS_WARNING || ae->old_status == RRDCALC_STATUS_CRITICAL)
+            ae->non_clear_duration += ae->duration;
+    }
+    else {
+        ae = target_alarm_entity->alarm_entry;
+        netdata_rwlock_wrlock(&host->health_log.alarm_log_rwlock);
+        // TODO: Update more necessary fields
+        ae->old_status = old_status;
+        ae->new_status = new_status;
+        netdata_rwlock_unlock(&host->health_log.alarm_log_rwlock);
+    }
 
     // link it
     netdata_rwlock_wrlock(&host->health_log.alarm_log_rwlock);
-    ae->next = host->health_log.alarms;
-    host->health_log.alarms = ae;
-    host->health_log.count++;
+    if(likely(!is_repeating)) {
+        ae->next = host->health_log.alarms;
+        host->health_log.alarms = ae;
+        host->health_log.count++;
+    }
+    else {
+        if(unlikely(!target_alarm_entity)) {
+            // TODO: Remove these entries once host is going to be freed
+            REPEATING_ALARM_ENTRY *rae = callocz(1, sizeof(REPEATING_ALARM_ENTRY));
+            rae->alarm_entry = ae;
+            if(unlikely(!host->health_rep_alarm_entry_list)) {
+                host->health_rep_alarm_entry_list = rae;
+            }
+            else {
+                rae->next = host->health_rep_alarm_entry_list;
+                host->health_rep_alarm_entry_list = rae;
```

Comment:
Why not just keep these two lines? The If isn't necessary. Exactly the same thing as lines 440-441 should be enough, rae->next will get a null if we fall into the case of lines 449-450 anyway.

Reply:
OK, I'll check that.

Reply:
These codes are not longer part of the change.